clang/test/SemaTemplate/template-id-expr.cpp

```diff
@@ -186,3 +186,74 @@ class E {
 #endif
 template<typename T> using D = int; // expected-note {{declared here}} 
 E<D> ed; // expected-note {{instantiation of}}
+
+namespace non_functions {
+
+#if __cplusplus >= 201103L
+namespace PR88832 {
+template <typename T> struct O {
+  static const T v = 0;
+};
+
+struct P {
+  template <typename T> using I = typename O<T>::v; // #TypeAlias
+};
+
+struct Q {
+  template <typename T> int foo() {
+    return T::template I<int>; // expected-error {{'P::I' is expected to be a non-type template, but instantiated to a type alias template}}
+    // expected-note@#TypeAlias {{type alias template declared here}}
```

Comment:
I think this is more of a preference - subjectively, I feel like using markers makes the diagnostics concentrated, and one can see them in a somewhat consistent block without guessing what else notes/warnings an error diagnostic would also provide, though this looks a bit untidy. :(

Reply:
> Can you add some more tests that involve an expression with this new type as the argument to a function call?

Done. Thanks for the feedback.

Reply:
I VASTLY prefer the markers.  It makes it easier to determine which error is causing the note.

Reply:
>  It makes it easier to determine which error is causing the note.

Iirc last time this came up in a discussion the consensus was that it was better not to use markers, but I think that was more about actual errors than notes, because this makes sense, yeah.

Reply:
Yeah, it doesn't make sense for errors/warnings: Those should be "right next" to the thing that caused them (so I prefer the  +/- Numeral for those).  However, with Notes, it makes a TON of sense, and IMO, should nearly ALWAYS be used.

Without them, we end up getting 'errors' and their notes are completely unintelligible, because they are spread out around the file and in no reasonable order.  It makes it so figuring out "this is an error only because of these instantiations" is impossible.

Reply:
That’s true; I’ve run into cases before where the notes just ended up being a mess...

Reply:
FWIW, I think I'm the 'first' person who has really been pushing for using the bookmarks for notes, no one ever has before, but when reviewing template stuff in particular, it is REALLY bad, so as code-owner, its one of the things I push for in new commits :)  

So you'll see very few that do it, but hopefully more happening.